compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs

```diff
@@ -52,16 +49,31 @@ impl<'tcx> FunctionCoverage<'tcx> {
         is_used: bool,
     ) -> Self {
         let num_counters = function_coverage_info.num_counters;
-        let num_expressions = function_coverage_info.num_expressions;
+        let num_expressions = function_coverage_info.expressions.len();
         debug!(
             "FunctionCoverage::create(instance={instance:?}) has \
             num_counters={num_counters}, num_expressions={num_expressions}, is_used={is_used}"
         );
+
+        // Prepare the subset of expression IDs that are directly used by one or
+        // more mappings, and should therefore have a corresponding `ExpressionUsed`
+        // statement in MIR.
+        //
+        // During MIR traversal, each `ExpressionUsed` statement we encounter will
+        // have its ID removed from this set. After traversal, any IDs remaining in
+        // the set must be ones that had their statements removed by MIR optimizations.
+        let mut expressions_expected_but_not_seen = BitSet::new_empty(num_expressions);
+        for Mapping { term, .. } in &function_coverage_info.mappings {
+            if let &CovTerm::Expression(id) = term {
+                expressions_expected_but_not_seen.insert(id);
+            }
+        }
```

Comment:
If an expression is only used indirectly, we automatically mark it as “seen” when initializing the set. If it ends up being unused, LLVM will delete it for us.

Reply:
To clarify a bit more:

- If a directly-used expression is detected as having been removed by MIR opts, then all mappings and expression operands referring to that expression will be replaced with zero.
- This can cause indirect-only expressions to become completely unused. We still emit them to LLVM, but LLVM has its own unused-expression pass that deletes them.